sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+In this document, I will go into more detail for each of these points as well as supply a proposal for how to address them. 
+
+It is important to note that this refactor will help us in our effort to decouple the CLI logic from the SDK. That way, when we are ready and decide to make that separation, we can do so more easily.
+
+
+
+### Current SDK implementation
+
+#### Flow
+
+The below image details the calls from when a user initalizes the *current* Client to when they receive a response.
+
+![CurrentSDKFlow](design/assets/CurrentSDKFlow.png)
+
+At a high level,
+
+1. The user initializes the *Client* with details needed to connect to the *Conjur server*
+
+    `client = conjur.Client(url="https://someserver", account="cucumber", password="123", ...)`.
+
+2. The *Client* constructor is called, ensuring all configuration and credential information needed to make a request have been passed in. If not, it fetches the missing information in a discovery-like manner. For example, for `login`, if a Keystore is not available, it will attempt to fetch credentials from the netrc. At this point the *API* is also initalized.
+
+3. *Client* is now initalized and the user can now make a request.
+
+   `list_values = client.list()`
+
+4. Once a request is made, the *Client* makes a call to *API* which handles the building of the request, merging parameters together, understanding which endpoint to use, etc.
+
+5. The *API* calls the *HTTPWrapper* which escapes the parameters and invokes the endpoint to the *Conjur Server*.
+
+6. A response is returned to the *HTTPWrapper*, handing it back to the calling *API* which formats the response to the desired output.
+
+7. *API* returns the formatted response to the *Client* and the *Client* returns it to the user. 
+
+
+
+### Limitations in current SDK Flow
+
+This section will be divided according to the three areas in the code that need the most attention (Client, Api, and Commands). In each section, I will outline their limitations, the consequences of their limitations, and provide a proposal for how to address them.
+
+##### Client constructor
+
+The Client constructor resembles the following:
+
+```python
+# client.py
+class Client():
+    def __init__(self,
+                 account: str = None,
+                 api_key: str = None,
+                 ca_bundle: str = None,
+                 debug: bool = False,
+                 http_debug=False,
+                 login_id: str = None,
+                 password: str = None,
+                 ssl_verify: bool = True,
+                 url: str = None):
+      ...
+      if not url or not login_id or (not password and not api_key):
+        # load in Configuration details (if not provided)
+        # load in Credential details (if not provided)
+        # instantiate API objects according to the different params provided
+        ...
+        if api_key:
+          ...
+        elif password:
```

Comment:
We are considering just one of those two arguments but not saying which we accepted, for example:
`client(api_key:"123", password: "123")`
will work, but no indication for client what is redundant, I don't say that we need to give back indication, but to be sure if we want to create a distinction between api_key and password arguments. 

Also, what will be the behavior when debug if False but http_debug is True ? 
I would also like to think about the order of those argument input as debug/http_debug should be at the end of the list

Reply:
IMO we should make the distinction between password and API key because with password, we need to perform a login to get the API key. 

> We are considering just one of those two arguments but not saying which we accepted

We can definitely document precedence in our docs

> Also, what will be the behavior when debug if False but http_debug is True ?
debug = true, see verbose logging of the process flow. debug = false, you won't see any logs at all printed to STDOUT. Only error or success. 

http_debug = true, we can thoroughly debug HTTP requests and HTTP responses